torch/_dynamo/guards.py

```diff
@@ -611,6 +642,145 @@ def _produce_guard_code(
+        ast.Index,
+        ast.Expr,
+        ast.Module,
+        # Leaf Expression nodes
+        ast.Name,
+        ast.Constant,
+        # Expr-Context nodes
+        ast.Load,
+        ast.Store,
+        ast.Del,
+        # Bool Operation nodes
+        ast.And,
+        ast.Or,
+        # Arithmetic Operation nodes
+        ast.Add,
+        ast.Sub,
+        ast.Mult,
+        ast.MatMult,
+        ast.Div,
+        ast.Mod,
+        ast.Pow,
+        ast.LShift,
+        ast.RShift,
+        ast.BitOr,
+        ast.BitXor,
+        ast.BitAnd,
+        ast.FloorDiv,
+        # Unary Operation nodes
+        ast.Invert,
+        ast.Not,
+        ast.UAdd,
+        ast.USub,
+        # Compare Operation nodes
+        ast.Eq,
+        ast.NotEq,
+        ast.Lt,
+        ast.LtE,
+        ast.Gt,
+        ast.GtE,
+        ast.Is,
+        ast.IsNot,
+        ast.In,
+        ast.NotIn,
+    )
+
+    class CSEVisitor(ast.NodeVisitor):
+        IGNORE = object()
+
+        def __init__(self) -> None:
+            self._expr_counter: Dict[str, int] = collections.defaultdict(lambda: 0)
+
+        def visit(self, node: ast.AST) -> Any:
+            if _ast_unparse_implemented(node) and not isinstance(
+                node, PyExprCSEPass.IGNORED_NODE_TYPES
+            ):
+                self._expr_counter[_ast_unparse(node)] += 1
+            super().visit(node)
+
+    class CSETransformer(ast.NodeTransformer):
+        def __init__(
+            self,
+            expr_counter: Dict[str, int],
+            preface: List[str],
+            gen_name: Callable[[], str],
+        ) -> None:
+            super().__init__()
+            self._expr_counter = expr_counter
+            self._preface = preface
+            self._gen_name = gen_name
+            self._expr_to_name: Dict[str, str] = {}
+
+        def visit(self, node: ast.AST) -> Any:
+            if _ast_unparse_implemented(node) and not isinstance(
+                node, PyExprCSEPass.IGNORED_NODE_TYPES
+            ):
+                expr = _ast_unparse(node)
+
+                # Replacement only occurs if a given expression is used more
+                # than once.
+                if self._expr_counter[expr] > 1:
+                    if expr not in self._expr_to_name:
+                        # Parent 'visit' is called so that we CSE the inner expressions first.
+                        #
+                        # The resulting expression is used as right-hand-side of the variable
+                        # assignment. i.e. we are CSE-ing the children before the parents.
+                        #
+                        # Indexing still uses the old 'node', since that's what was counted
+                        # by the 'NodeVisitor'.
+                        node_ = super().visit(node)
+                        expr_ = _ast_unparse(node_)
+                        var_name = self._gen_name()
+                        self._preface.append(f"{var_name} = {expr_}")
```

Comment:
Hoisting the expressions all the way to the beginning is not sound in general, and I think this explains the need for the try...catch.

The order we run guards matters. Suppose that we have a guard on x being tensor, and another guard on x.grad. We must test that x is tensor, before we attempt to access x.grad. With the hoisting here, you may violate the dependencies on guard accesses.

Reply:
> I think this explains the need for the try...catch.
Yes.